core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/ContainerPool.scala

```diff
@@ -303,26 +326,94 @@ class ContainerPool(childFactory: ActorRefFactory => ActorRef,
+      val desiredCount: Int =
+        if (init) config.initialCount
+        else {
+          if (scheduled) {
+            // scheduled/reactive config backfill
+            config.reactive
+              .map(c => getReactiveCold(c, kind, memory).getOrElse(c.minCount)) //reactive -> desired is either cold start driven, or minCount
+              .getOrElse(config.initialCount) //not reactive -> desired is always initial count
+          } else {
+            // normal backfill after removal - make sure at least minCount or initialCount is started
+            config.reactive.map(_.minCount).getOrElse(config.initialCount)
+          }
+        }
+
+      logging.info(
+        this,
+        s"[kind: ${kind} memory: ${memory.toString}] needs ${desiredCount} desired prewarmed container")
+
+      // remove expired
+      config.reactive.foreach { _ =>
+        val expiredPrewarmedContainer = prewarmedPool
+          .filter { warmInfo =>
+            warmInfo match {
+              case (_, p @ PreWarmedData(_, `kind`, `memory`, _, _)) if p.isExpired() => true
+              case _                                                                  => false
+            }
+          }
+        // emit expired container counter metric with memory + kind
+        val removedCount = expiredPrewarmedContainer.size
+        MetricEmitter.emitHistogramMetric(LoggingMarkers.CONTAINER_POOL_EXPIRED(memory.toString, kind), removedCount)
         logging.info(
           this,
-          s"found ${currentCount} started and ${startingCount} starting; ${if (init) "initing" else "backfilling"} ${config.count - containerCount} pre-warms to desired count: ${config.count} for kind:${config.exec.kind} mem:${config.memoryLimit.toString}")(
+          s"[kind: ${kind} memory: ${memory.toString}] removed ${removedCount} expired prewarmed container")
+        expiredPrewarmedContainer.map(_._1 ! Remove)
+      }
+
+      if (currentCount < desiredCount) {
+        logging.info(
+          this,
+          s"found ${currentCount} started and ${startingCount} starting; ${if (init) "initing" else "backfilling"} ${desiredCount - currentCount} pre-warms to desired count: ${desiredCount} for kind:${config.exec.kind} mem:${config.memoryLimit.toString}")(
           TransactionId.invokerWarmup)
-        (containerCount until config.count).foreach { _ =>
-          prewarmContainer(config.exec, config.memoryLimit)
+        (currentCount until desiredCount).foreach { _ =>
+          prewarmContainer(config.exec, config.memoryLimit, config.reactive.map(_.ttl))
         }
       }
     }
+    if (scheduled) {
+      // emit cold start counter metric with memory + kind
+      coldStartCount foreach { coldStart =>
+        val coldStartKey = coldStart._1
+        val coldStartValue = coldStart._2
+        MetricEmitter.emitHistogramMetric(
+          LoggingMarkers.CONTAINER_POOL_COLDSTART(coldStartKey.memory.toString, coldStartKey.kind),
+          coldStartValue)
```

Comment:
I would name this metric "prewarmColdstart" - to not confuse it with containerStart (which has a tag for containerState). These values are not just cold starts - but cold starts that matched prewarm configs (and did not have a prewarm available)

Reply:
updated accordingly